File: CPP/explore/MaxProfit.cpp
#include <vector>

class Solution
{
public:
  int maxProfit(std::vector<int> &prices)
  {
    int profit = 0;

    if (prices.size() == 0 || prices.size() == 1)
      return profit;

    for (int i = 0; i < prices.size() - 1; i++)
    {
      if (prices[i] < prices[i + 1])
      {
        profit += prices[i + 1] - prices[i];
      }
    }

    return profit;
  }
};

File: JS/old_solutions/MaxProfit.js
/**
 * Given array of a the stock price for each day, find the max profit for all the days.
 * @param {number[]} prices
 * @return {number}
 */
const maxProfit = function(prices) {
	let difference = 0;
	if (prices.length > 0) {
		prices.reduce((accumulator, current) => {
			if (current > accumulator) {
				difference += current - accumulator;
			}
			return current;
		});
	}

	return difference;
};

// No initial value is set for accumulator, so it takes the first element and current takes the second.
// This is a two pointers problem, but it turns out you only need to worry about next.

const maxProfit = function(prices) {
	let profit = 0;

	if (prices.length > 1) {
		for (let i = 0; i < prices.length - 1; i++) {
			if (prices[i + 1] > prices[i]) {
				profit += prices[i + 1] - prices[i];
			}
		}
	}

	return profit;
};
